sqlx-core/src/sqlite/connection/establish.rs

```diff
@@ -131,6 +163,57 @@ impl EstablishParams {
             sqlite3_extended_result_codes(handle.as_ptr(), 1);
         }
 
+        if !self.extensions.is_empty() {
+            // Enable extension loading via the db_config function, as recommended by the docs rather
+            // than the more obvious `sqlite3_enable_load_extension`
+            // https://www.sqlite.org/c3ref/db_config.html
+            // https://www.sqlite.org/c3ref/c_dbconfig_defensive.html#sqlitedbconfigenableloadextension
+            status = unsafe {
+                sqlite3_db_config(
+                    handle.as_ptr(),
+                    SQLITE_DBCONFIG_ENABLE_LOAD_EXTENSION,
+                    1, // '1' here means enable the C API function but not the SQL function
+                    null::<i32>(),
+                )
+            };
+            if status != SQLITE_OK {
+                return Err(Error::Database(Box::new(SqliteError::new(handle.as_ptr()))));
+            }
+
+            for ext in self.extensions.iter() {
+                // `sqlite3_load_extension` is unusual as it returns its errors via an out-pointer
+                // rather than by calling `sqlite3_errmsg`
+                let mut error = null_mut();
+                status = unsafe {
+                    sqlite3_load_extension(
+                        handle.as_ptr(),
+                        ext.0.as_ptr(),
+                        ext.1.as_ref().map_or(null(), |e| e.as_ptr()),
+                        addr_of_mut!(error),
+                    )
+                };
+
+                if status != SQLITE_OK {
+                    // SAFETY: We become responsible for any memory allocation at `&error`, so test
+                    // for null and take an RAII version for returns
+                    let err_msg = if !error.is_null() {
+                        unsafe {
+                            let e = CStr::from_ptr(error).into();
+                            sqlite3_free(error as *mut c_void);
+                            e
+                        }
+                    } else {
+                        CString::new("Unknown error when loading extension")
+                            .expect("text should be representable as a CString")
+                    };
+                    return Err(Error::Database(Box::new(SqliteError::extension(
+                        handle.as_ptr(),
+                        &err_msg,
+                    ))));
+                }
+            }
+        }
```

Comment:
For sanity reasons, this should probably set `SQLITE_DBCONFIG_ENABLE_LOAD_EXTENSION` back to `0` after the extensions have been loaded.

Without that, I could see someone coming along, seeing this and saying "oh SQLx leaves extension loading enabled after startup, that's a security vulnerability," even if it's technically not since using the C API would be required.

Reply:
Done, and since we're calling it twice I refactored the mysterious int-flags into a function with a descriptive enum as an argument.